evals/elsuite/ecn.py

```diff
@@ -0,0 +1,102 @@
+import random
+import re
+
+import evals
+import evals.metrics
+from evals import CompletionFn
+
+
+class Ecn(evals.Eval):
+    def __init__(
+        self,
+        completion_fns: list[CompletionFn],
+        samples_jsonl: str,
+        num_few_shot: int = 0,
+        few_shot_jsonl: str = None,
+        **kwargs
+    ):
+        super().__init__(completion_fns, **kwargs)
+        self.samples_jsonl = samples_jsonl
+        self.num_few_shot = num_few_shot
+        self.few_shot_jsonl = few_shot_jsonl
+
+        if self.num_few_shot > 0:
+            assert few_shot_jsonl is not None, "few shot requires few shot sample dataset"
+            self.few_shot_jsonl = few_shot_jsonl
+            self.few_shot = evals.get_jsonl(self.few_shot_jsonl)
+
+    def run(self, recorder):
+        """
+        Called by the `oaieval` CLI to run the eval. The `eval_all_samples` method calls `eval_sample`.
+        """
+
+        samples = self.get_samples()
+        self.eval_all_samples(recorder, samples)
+        events = recorder.get_events("match")
+        num_events = len(list(events))
+
+        total_score = sum(map(lambda e: e.data["score"], events))
+        average_score = total_score / num_events
+        valid_format = f"{round(sum(map(lambda e: e.data['valid_format'], events)) / num_events * 100, 2)}%"
+
+        return {
+            "accuracy": evals.metrics.get_accuracy(events),
+            "total_score": total_score,
+            "average_score": average_score,
+            "valid_format": valid_format,
+        }
+
+    def eval_sample(self, sample, rng: random.Random):
+        """
+        Called by the `eval_all_samples` method to evaluate a single sample.
+
+        ARGS
+        ====
+        `test_sample`: a line from the JSONL test file
+        `rng`: should be used for any randomness that is needed during evaluation
+
+        This method does the following:
+        1. Generate a prompt that contains the task statement, a few examples, and the test question.
+        2. Generate a completion from the model.
+        2. Check if the generated answer is correct.
+        """
+
+        prompt = sample["input"]
+        expected = sample["ideal"].strip()
+        if self.num_few_shot > 0:
+            prompt = sample["input"][:-1]
+            for s in self.few_shot[: self.num_few_shot]:
+                prompt += s["sample"]
+            prompt += sample["input"][-1:]
+
+        result = self.completion_fn(
+            prompt=prompt,
+        )
+        sampled = result.get_completions()[0].strip()
+
+        valid_format = re.fullmatch(r'(^[A-E]$)|(^[A-E](,[A-E]){0,4}$)', sampled) is not None
+
+        match = sampled == expected
```

Comment:
Do we ever have `C,B,E` as an answer from the LLM? Seems unlikely, but just wondering if you checked.
Do we store the completions btw in order to debug / understand?

Reply:
Ok, we should probably "just" set match to `True` in the logic below actually.

Reply:
> Do we store the completions btw in order to debug / understand?

I added the logs to the PR. Yes, GPT-4 follows the expected output perfectly, and GPT-3.5 does a good job too.

> we should probably "just" set match to True in the logic below.

I'm going to change match to `score == 1.0`, as that way we handle the slight format issues (while also recording format correctness).

Reply:
[1ab986a](https://github.com/alan-eu/evals/pull/1/commits/1ab986a7709a8525f559cd0647ca24eaba023b8e)